File: diemBFT/src/main/java/diem/component/BlockTree.java
package diem.component;

import diem.model.Block;
import diem.model.QuorumCertificate;
import diem.model.VoteMessage;
import diem.model.*;

import java.util.List;

public class BlockTree {

    private List<Block> pendingBlockTree;
    private List<VoteMessage> pendingVotes;
    private QuorumCertificate highQC;
    static public QuorumCertificate highCommitQC;


    public void processQC(QuorumCertificate qc) {
        if (qc.ledgerCommitInfo.commitStateId != null) {
            Ledger.commit(qc.voteInfo.parentId);
            // TODO Pending Block tree
//            pendingBlockTree.
            highCommitQC = getMax(qc, highCommitQC);
        }
        highQC = getMax(qc, highQC);
    }

    // Revisit
    public void executeAndInsert(Block block) {
        Ledger.speculate(block.quorumCertificate.blockId, block.id, block.payload);
        pendingBlockTree.add(block);
    }

    public QuorumCertificate processVote(VoteMessage vote) {
        processQC(vote.highCommitQC);
        //TODO Implement this code
//        if (pendingVotes == )
        QuorumCertificate qc = null;
        return qc;
    }

    public Block generateBlock(List<Transaction> transaction, int currentRound) {
        return new Block(null, 0, transaction, highQC, hashCode());
    }


    private static QuorumCertificate getMax(QuorumCertificate qc, QuorumCertificate highCommitQC) {
        return null;
    }

}

File: diemBFT/test2/blocktree.py
# from ledger import Ledger
#
# class BlockTree:
#     def __init__(self, voteInfo, ledgerCommitInfo, voteMsg, qc, block,pendingBlockTree,pendingVotes,highQC,highCommitQC):
#         self.voteInfo = voteInfo
#         self.ledgerCommitInfo = ledgerCommitInfo
#         self.voteMsg = voteMsg
#         self.qc = qc
#         self.block = block
#         self.pendingBlockTree = pendingBlockTree
#         self.pendingVotes = pendingVotes
#         self.highQC = highQC
#         self.highCommitQC = highCommitQC
#
#     def process_qc(qc):
#         if qc.ledgerCommitInfo.commitStateId is not None:
#             Ledger.commit(qc.vote_info.parent_id)
#             pending_block_tree.prune(qc.vote_info.parent_id)
#             high_commit_qc = maxround {qc, high commit qc}
#         high_qc = maxround{qc, high qc}
#
#     def execute_and_insert(b):
#         Ledger.speculate(b.qc.block id, b.id, b.payload)
#         pending_block_tree.add(b)